samples/switchless/README.md

```diff
@@ -0,0 +1,133 @@
+
+## How does OE support switchless OCALLs
+
+OE only supports synchronous switchless OCALLs currently. When the caller within an enclave makes a switchless OCALL, the trusted OE runtime creates a `job` out of the function call. The `job` object includes information such as the function ID, the parameters marshaled into a buffer, and a buffer for holding the return value(s). The job is posted to a shared memory region which both the enclave and the host can access.
+
+A host worker thread checks and retrieves `job` from the shared memory region. It uses the untrusted OE runtime to process the `job` by unmarshaling the parameters, then dispatching to the callee function, and finally relaying the result back to the trusted OE runtime, which is further forwarded back to the caller.
+
+To support simultaneous switchless OCALLs made from enclaves, the host workers are multi-threaded. OE allows users to configure how many host worker threads are to be created for servicing switchless OCALLs. The following example illustrates how to do that. A word of caution is that too many host worker threads might increase competition of cores between threads and degrade the performance. Therefore, if a enclave has switchless calls enabled, OE caps the number of host worker threads for it to the number of enclave threads specified.
+
+## About the EDL
+
+First we need to define the functions we want to call between the host and the enclave. To do this we create a `switchless.edl` file:
+
+```edl
+enclave {
+    trusted {
+        public void enclave_helloworld();
+
+    };
+
+    untrusted {
+        void host_helloworld();
+        void host_helloworld_switchless() transition_using_threads;
+    };
+};
+```
+
+Function `host_helloworld_switchless`'s declaration ends with keyword `transition_using_threads`, indicating it should be called switchlessly at run time. However, this a best-effort directive. OE runtime may still choose to fall back to a tradition OCALL if switchless call resources are unavailable, e.g., the enclave is not configured as switchless-capable, or the host worker threads are busy servicing other switchless OCALLs.
+
+To generate the functions with the marshaling code, the `oeedger8r` tool is called in both the host and enclave directories from their Makefiles. For example:
+
+```bash
+cd host
+oeedger8r ../switchless.edl --untrusted --experimental
+```
+
+`oeedger8r` needs the command line flag `--experimental` to be able to recognize the keyword `transition_using_threads`.
+
+## About the host
+
+The host first defines a structure specifically for configuring switchless calls. In this case, we specify the first field `2` as the number of host worker threads for switchless OCALLs. The 2nd field specifies the number of enclave threads for switchless ECALLs. Since switchless ECALL is not yet implementated, we require the 2nd field to be `0`.
+
+```c
+oe_enclave_config_context_switchless_t config = {2, 0};
```

Comment:
I think hard coding the number is almost always guaranteed to be wrong.   I don't think a sample should recommend bad practice, i.e., hard coding numbers that are bad.   So either explain why "2" is the best number (in which case lots of apps can just copy this sample code and hard code the same number), or remove the number 2 from the sample.  E.g., make it be a command line argument.

Reply:
Added the logic behind '2' (it's 1 now actually).